src/services/identity/UsersService.ts

```diff
@@ -188,26 +210,120 @@ class UsersService {
   }
 
   async sendEmailOtp(email: string) {
-    const otp = this.otp.generate(email)
-    const expiry = this.otp.getExpiryMinutes()
+    const { otp, hashedOtp } = await this.otpService.generateLoginOtpWithHash()
+
+    // Reset attempts to login
+    const otpEntry = await this.otpRepository.findOne({ where: { email } })
+    if (!otpEntry) {
+      // create new entry
+      await this.createOtpEntry(email, OtpType.Email, hashedOtp)
+    } else {
+      await otpEntry?.update({
+        hashedOtp,
+        attempts: 0,
+        expiresAt: this.getOtpExpiry(),
+      })
+    }
 
     const subject = "One-Time Password (OTP) for IsomerCMS"
-    const html = `<p>Your OTP is <b>${otp}</b>. It will expire in ${expiry} minutes. Please use this to verify your email address.</p>
+    const html = `<p>Your OTP is <b>${otp}</b>. It will expire in ${milliSecondsToMinutes(
+      PARSED_EXPIRY
+    )} minutes. Please use this to verify your email address.</p>
     <p>If your OTP does not work, please request for a new OTP.</p>
     <p>IsomerCMS Support Team</p>`
     await this.mailer.sendMail(email, subject, html)
   }
 
   async sendSmsOtp(mobileNumber: string) {
-    const otp = this.otp.generate(mobileNumber)
-    const expiry = this.otp.getExpiryMinutes()
+    const { otp, hashedOtp } = await this.otpService.generateLoginOtpWithHash()
+
+    // Reset attempts to login
+    const otpEntry = await this.otpRepository.findOne({
+      where: { mobileNumber },
+    })
+    if (!otpEntry) {
+      await this.createOtpEntry(mobileNumber, OtpType.Mobile, hashedOtp)
+    } else {
+      await otpEntry?.update({ hashedOtp, attempts: 0 })
+    }
 
-    const message = `Your OTP is ${otp}. It will expire in ${expiry} minutes. Please use this to verify your mobile number`
+    const message = `Your OTP is ${otp}. It will expire in ${milliSecondsToMinutes(
+      PARSED_EXPIRY
+    )} minutes. Please use this to verify your mobile number`
     await this.smsClient.sendSms(mobileNumber, message)
   }
 
-  verifyOtp(value: string, otp: string) {
-    return this.otp.verify(value, otp)
+  private async verifyOtp(otpEntry: Otp | null, otp: string) {
+    // TODO: Change all the following to use AuthError after FE fix
+    if (!otp || otp === "") {
+      throw new BadRequestError("Empty OTP provided")
+    }
+
+    if (!otpEntry) {
+      throw new BadRequestError("OTP not found")
+    }
+
+    if (otpEntry.attempts >= PARSED_MAX_NUM_OTP_ATTEMPTS) {
+      throw new BadRequestError("Max number of attempts reached")
+    }
+
+    if (!otpEntry?.hashedOtp) {
+      await otpEntry.destroy()
+      throw new BadRequestError("Hashed OTP not found")
+    }
+
+    // increment attempts
+    await otpEntry.update({ attempts: otpEntry.attempts + 1 })
```

Comment:
not a blocker but just a site note that we don't increment attempts when the submitted otp is blank

Reply:
decided together that this was acceptable for our use cases with rate limiter in place